Assorted changes and a couple of fixes to the fancy sheets

PJBrs opened the pull request:
Hi, I just found dungeon-sheets and I like it very much! I had been writing my character sheets by hand in latex using MSavage's style, but that isn't very automated.

I hope to contribute a little bit with a couple of ideas that are focused on the features pages. I think basically the patches speak for themselves, and for now I'd mainly like to get some reactions as to whether these patches are actually wanted, and perhaps what other / further directions to take with this.

I can also supply an example of the type of layout that I'm trying to achieve. I'm keeping a small wishlist of my own, with items like:

- Ranger: Add feature options for Favored Enemy and Natural Explorer.
- XP: Add Milestone option.
- Allow handwritten boxes for rations, torches, arrows, bullets.
- Add versatile damage automatically
- Darkvision is always listed as source: race, but it can also be given by a (sub)class

I should add that I can code only a little bit, but I tested the changes on all the example sheets.

Anyway, I really hope to hear whether this type of work is appreciated!

bw-mutley commented:
> Hi, I just found dungeon-sheets and I like it very much! I had been writing my character sheets by hand in latex using MSavage's style, but that isn't very automated.
> 
 Hi there, @PJBrs ! 
About this part, did you try using MSavage from Dungeonsheets? I've contributed to this part, let me know of any issues I can help.

Regards,
Mauricio

PJBrs commented:
Hi @bw-mutley!

Thanks! I think most of the more structural changes apply to the features pages, hardly any to MSavaga's style. I've tested both though. 

If you have any suggestions though for continuing this PR, please let me know. In the meantime, I've added one more substantial change that renders the monster blocks more in keeping with the RPGTex D&D style. Problem is, my repo now is one long list of assorted changes, including spelling fixes, small additions, small enhancements and substantial formatting changes that affect how everything looks... 

I'm force-updating the PR now, and then I'll see whether I can sort it better. In the meantime, any suggestion at all would be appreciated!

PJBrs commented:
So, I sorted all my patches. The first five patches are trivial spelling and whitespace fixes.
The next four add some equipment and offer very small corrections:
- Ice Knife: put the material components in the right place
- Armor: Add shield weight
- Equipment: Add plural of book
- Add potion of greater healing to equipment catalogue

The "Fix" pdfbookmarks patch does not fully work, but as far as I can tell, it's less error-prone than the current behaviour.

The next five patches all represent small enhancements:
- Start page number of feature pages at 4
- Distinguish between racial and class darkvision
- Proficiencies: Harmonize capitalization and punction in MSavage.tex
- Add 3rd and 5th level spells for Drow Magic feat
- Add spellcasting details to latex spellbook form

All the remaining patches offer rather substantial changes that I personally like, and I hope for some feedback on them.

These two patches sort spells by name and level and add headings for each spell level in the features pages:
- Sort spells by name and level
- Add subheaders for spell level in the pdf

These two patches sort feats by origin and level (character feats, class feats, background feats and racial feats) and add associated headers in the feature pages:
- List feats in the order that they're added
- Separate headers for Character, Class, Racial and Background Feats 

I tried to write these trying not to be invasive and ended up doing quite a lot of pattern matching in latex. Could have done them simpler.

The final three patches all concern relatively minor aesthetic changes that I liked:
- List versatile damage in brackets if it exists <-- This lists versatile damage on the character sheet if a weapon deals versatile damage.
- MSavagetemplate: set XP to milestone if -1 <-- This enables users to have Milestone XP on their character sheet by entering -1 XP in their character .py-file.
- Use all RPGtex options for monster formatting <-- This is actually quite substantial and tries to make monster box formatting to be more true to RPG-tex.

Basically, I'm looking for comments on all of these, regarding:
- Which of these changes, if any, are wanted at all?
- For those changes that are wanted, how should they be implemented?
- How should I go about these patches anyway?

bw-mutley commented:
> Problem is, my repo now is one long list of assorted changes, including spelling fixes, small additions, small enhancements and substantial formatting changes that affect how everything looks...

exactly how my local files are at the moment. I've made several additions and corrections too, but can't keep track of everything in commits like you did - seens like I have a lot to learn on how to keep with people here. To properly contribute again, I will - at some point - have to dowload the master branch again. I've seen many issues in my local files which were probably fixed.

canismarko commented:
Hi, @PJBrs.

Apologies for the slow response. My day job got unexpectedly chaotic. Thank you for all the hard work and for organizing these commit in a useful way. I will take a closer look when things calm down a little bit and I get my feet back underneath me.

PJBrs commented:
Hi @canismarko , take your time! By now it should be clear that I couldn't help myself anyway :-)

I think, given all patches I now added, the majority should be sufficiently palatable, but others also represent design choices that you might disagree with, and then there's the matter of whether my implementations are agreeable, aside from their effects.

I hope things will calm down soon!

PJBrs commented:
@canismarko I noticed you triggered some tests. The latex ones fail because of the changes I made. The below patch adapts the tests to my other patches (and makes them pass). I'll look into the other failing tests another time.

```
--- a/tests/test_make_sheets.py
+++ b/tests/test_make_sheets.py
@@ -235,7 +235,7 @@ class TexCreatorTestCase(unittest.TestCase):
     def test_create_features_tex(self):
         char = self.new_character()
         tex = make_sheets.create_features_content(character=char, content_suffix="tex")
-        self.assertIn(r"\section*{Features}", tex)
+        self.assertIn(r"\section*{Class Features}", tex)
         self.assertIn(r"\subsection*{Martial Arts}", tex)
 
     def test_create_magic_items_tex(self):
@@ -248,7 +248,8 @@ class TexCreatorTestCase(unittest.TestCase):
         char = self.new_character()
         tex = make_sheets.create_spellbook_content(character=char, content_suffix="tex")
         self.assertIn(r"\section*{Spells}", tex)
-        self.assertIn(r"\section*{Invisibility}", tex)
+        self.assertIn(r"\subsection*{2nd-Level Spells}", tex)
+        self.assertIn(r"\subsubsection*{Invisibility}", tex)
 
     def test_create_infusions_tex(self):
         char = self.new_character()

```

PJBrs commented:
I added the test I posted previously in the comments. Also, I reordered the new patches a bit to leave the DND-monster formatting at the end. I tested all monsters I could, made many formatting fixes as well, only remaining thing is to add bulleted lists.

And I wrote one more patch to enable DND-tables throughout, after I conjured up a regex that translated docutil's supertabular table header to a dndtable header. Next step is probably to see if I can improve any table I find.

PJBrs commented:
@canismarko @bw-mutley 
I think I really have no other patches to add right now. I have put a character sheet online to demonstrate the effects of all the patches in this PR together, note the feat and spell headings, the dndtables with text wrapping in cells, and the monster box formatting (I've given this character, a druid, a sizeable amount of companions):
[https://floorenpj.nl/~pj/yordy.pdf](url)

@canismarko How might I make this PR easier to review? I can, for instance, make separate PRs for different groups of patches, for instance as follows:

1. Spelling fixes / bug fixes
2. MSavage style sheet enhancements
3. Other small enhancements
4. Monster formatting
5. DNDtables
6. Spellbook headings per spell level
7. Separated character, class, racial, and background feats

@bw-mutley Perhaps you'd be interested to take a look at the various patches that deal specifically with the MSavage style sheet?

Finally, I tried to run the dungeonsheets tests, which complain about some missing html pages, but the latex tests complete fine:

```
$ pytest3 tests
============================================================================================== test session starts ==============================================================================================
platform linux -- Python 3.9.16, pytest-7.2.1, pluggy-1.0.0
rootdir: /home/pj/Src/dungeon-sheets/dungeon-sheets
plugins: cov-3.0.0
collected 125 items                                                                                                                                                                                             

tests/test_character.py ........................                                                                                                                                                          [ 19%]
tests/test_content_registry.py .....                                                                                                                                                                      [ 23%]
tests/test_dice.py ......                                                                                                                                                                                 [ 28%]
tests/test_encounter.py ..                                                                                                                                                                                [ 29%]
tests/test_epub.py ...                                                                                                                                                                                    [ 32%]
tests/test_equipment_reader.py .                                                                                                                                                                          [ 32%]
tests/test_features.py .......                                                                                                                                                                            [ 38%]
tests/test_html.py .s.........                                                                                                                                                                            [ 47%]
tests/test_latex.py .s.........                                                                                                                                                                           [ 56%]
tests/test_magic_items.py ..                                                                                                                                                                              [ 57%]
tests/test_make_sheets.py ...FFF.FFF...........                                                                                                                                                           [ 74%]
tests/test_monsters.py ....                                                                                                                                                                               [ 77%]
tests/test_multiclass.py ....                                                                                                                                                                             [ 80%]
tests/test_random_tables.py ..                                                                                                                                                                            [ 82%]
tests/test_readers.py .........                                                                                                                                                                           [ 89%]
tests/test_spells.py ....                                                                                                                                                                                 [ 92%]
tests/test_stats.py ......                                                                                                                                                                                [ 97%]
tests/test_weapon.py ...                                                                                                                                                                                  [100%]

[snip!]
============================================================================================ short test summary info ============================================================================================
FAILED tests/test_make_sheets.py::EpubOutputTestCase::test_character_file_created - jinja2.exceptions.TemplateNotFound: preamble.html
FAILED tests/test_make_sheets.py::EpubOutputTestCase::test_character_html_content - jinja2.exceptions.TemplateNotFound: preamble.html
FAILED tests/test_make_sheets.py::EpubOutputTestCase::test_gm_file_created - jinja2.exceptions.TemplateNotFound: preamble.html
FAILED tests/test_make_sheets.py::HtmlCreatorTestCase::test_create_extra_gm_content - jinja2.exceptions.TemplateNotFound: extra_gm_content.html
FAILED tests/test_make_sheets.py::HtmlCreatorTestCase::test_create_monsters_html - jinja2.exceptions.TemplateNotFound: monsters_template.html
FAILED tests/test_make_sheets.py::HtmlCreatorTestCase::test_create_party_summary_html - jinja2.exceptions.TemplateNotFound: party_summary_template.html
============================================================================= 6 failed, 117 passed, 2 skipped, 5 warnings in 38.05s =============================================================================
```

bw-mutley commented:
Hi there, @PJBrs , sorry for the late reply. 

> Perhaps you'd be interested to take a look at the various patches that deal specifically with the MSavage style sheet?

Thanks for the contribution. I will fork your branch and try to do it this week.

PJBrs commented:
@bw-mutley Great! 

Let me know if you need the MSavage style sheet changes in a separate branch...

bw-mutley commented on dungeonsheets/classes/sorceror.py:
In the 10th (2018) printed version of PHB, the class name is *sorcerer*, not sorceror. In dungeonsheets, however, it was written sorceror. I thought it would be the case to change the other references.

bw-mutley commented:
> @bw-mutley Great!
> 
> Let me know if you need the MSavage style sheet changes in a separate branch...

Hi there, PJ! Only now I could have some spare time to work on it. I've made a small comment about the sorcerer -> sorceror change, please take a look. About the MSavagesheet: I will need some more time since I've changed my PC and ended missing some tex files needed.

PJBrs commented:
Hi @bw-mutley I thought sorceror was the correct spelling, but indeed, not only the PHB but also standard dictionaries prefer sorcer_er_. I wonder whether we should _revert_ or _reverse_ that spelling patch ;)

bw-mutley commented:
> Hi @bw-mutley I thought sorceror was the correct spelling, but indeed, not only the PHB but also standard dictionaries prefer sorcer_er_. I wonder whether we should _revert_ or _reverse_ that spelling patch ;)

I suppose it is up to @canismarko to decide. But my suggestion is to push a PR fixing all 'sorceror' ocurrences into 'sorcerer'.

PJBrs commented:
I stashed that sorcer_er_ spelling fix for now. 

Unfortunately, I had some uncommitted local changes that have now been added to this PR, one to make more tables nice and beautiful, and a compile fix for Monster Slayer Ranger subclass... Sorry!

@canismarko Like I said, let me know if you want me to separate this PR into smaller, topically consistent PRs...

PJBrs commented:
I noticed there has been some activity in this repo again, and I bumped into some small bits while making a new character, and I decided to reorganise this pull request. Hopefully, the different patches, and how they fit together, make more sense now.

Everything after Latex feature pages is potentially rather invasive. I hope at least some of these are ripe for cherry-picking.

canismarko commented:
Sorry I haven't gotten to this in so long. It is a lot to look through all at once. Perhaps it would be better in a few different PRs? The spelling mistakes and stuff like that I can easily take. New content is great too.

I noticed there's a couple commits that change the order that features and spells. Maybe this should be a command line option? I can see why someone might want spells in level order, but I also can see why someone might want them in alpha order. I originally put them in alpha order because that's how I look them up when I want to know what they do.

Would it still be useful to you if it were ``--spell-order=level`` and ``--feature-order=natural`` or something like that?

PJBrs commented:
I can make a couple of separate pull requests, starting with three relatively un-invasive ones. A **first** one could simply be the spelling fixes and the four obvious bug fixes:

- Features: Sorceror: Fix "Favored by the Gods
- Features: Ranger: Fix HuntersSense feat, fixes Monster Slayer subclass compile
- Spells: Ice Knife: put the material components in the right place
- Examples: Paladin: great-weapon master --> great-weapon fighting
- Assorted spelling and whitespace fixes

I could make a **second** one with miscellaneous small additions:

- Equipment: Add potion of greater healing
- Equipment: Add plural of book
- Armor: Add shield weight

And I could make a **third** one that only does some text formatting changes:

- Formatting: Spells: Prestidigitation: Fix bullet list
- Formatting: Monsters: Assorted formatting fixes
- Formatting: Beautify tables

It would seem to me that these are the changes that ought to be relatively trivial to review :) I'd like to see that you make of these patches first, and then I can see how to submit the more invasive enhancements. To be sure, I can try to make that commandline option! And I'm not entirely happy with how I implemented ordering feats anyway. 

So, how does that sound?

canismarko commented:
That sounds wonderful, thank you for your patience. It will be great to finally get this all merged in.

PJBrs commented:
Hi @canismarko - I've just created three separate pull requests. They all apply to master, so I don't think it matters at all in which order you pull them. They are 11 patches in total. Hope this helps!

PJBrs commented:
Hi @canismarko , I rebased my entire local tree and force pushed it over here. Incidentally, I also noticed that there have been quite some updates in MSavage's dndtemplate itself, which required some additional work on our side (see https://github.com/canismarko/dungeon-sheets/pull/147/commits/e2ba58eb9bb5b25bb75c9bc311e942952e1bce5a).

I'd like to propose the following three steps:

1. See what you make of my open pull request here - https://github.com/canismarko/dungeon-sheets/pull/155
2. Open a bug report about image positioning in the MSavage latex template
3. Open an additional pull request with the two patches that enhance how features are applied)
4. Open an additional pull request for the now seven patches that touch upon the MSavage latex template

How does that sound?

@bw-mutley, if possible, I'd really like you to check those seven patches for the MSavage latex template. If you need it, I have a local ~/texmf tree that includes the dndtemplate and the rpgtex latex format; I could package that for you, if needs be.

bw-mutley commented:
> @bw-mutley, if possible, I'd really like you to check those seven patches for the MSavage latex template. If you need it, I have a local ~/texmf tree that includes the dndtemplate and the rpgtex latex format; I could package that for you, if needs be.

Hi there, @PJBrs ! Thanks for the awesome contribution. I've already read that and approve both the idea and implementation, just didn't test it yet. If you already tested, I am ok with that. If you want me to test, I will ask for you to wait until the weekend.

PJBrs commented:
Hi @bw-mutley that's nice to hear! I wouldn't like to clobber this repo with pull requests, but I did prepare a separate branch with only MSavage latex template fixes. And it would be nice, I think, for you to test it a little bit. I read in my other pulll request that you had some trouble with git. I suggest some, if not all, of the following:
```
git clone https://github.com/canismarko/dungeon-sheets/
cd dungeon-sheets/
remote add pjbrs https://github.com/PJBrs/dungeon-sheets
git fetch pjbrs
git checkout -B pjbrs/MSavage-latex-template

```

This will get you a clean repo with only my changes to MSavage's latex template. With testing, it might be easier for @canismarko to decide to pull these changes. I've made a character sheet that I displays the working of all the patches in this series. Compare pictures to see differences.

Character without MSavage latex template patches:
![image](https://github.com/canismarko/dungeon-sheets/assets/5177219/11a8d928-9f74-4644-832d-7b6544a7fa5d)

Character without MSavage latex template patches:
![image](https://github.com/canismarko/dungeon-sheets/assets/5177219/b18d8c81-df3f-483e-b84a-9c8d56764b65)

Here's the source (remove .txt from the filename):
[llaewin.py.txt](https://github.com/canismarko/dungeon-sheets/files/13835573/llaewin.py.txt)

bw-mutley commented:
hi there, PJ! Thanks for your instructions, I've followed and proceeded with tests here. But I think I'm still missing something, because I wasn't able to reproduce the sheets you've posted. Most noticiable difference is the attributes fields, which is shown with the bonuses in the larger space in your example but not in mine. Checking further, I found the template in my local files are not the same as you have in your branch here:

https://github.com/PJBrs/dungeon-sheets/blob/MSavage-latex-template/dungeonsheets/forms/MSavage_template.tex

Could you help me to figure what is wrong in our process? BTW, reading the MSavage_template.tex above I would guess it wouldn't reproduce the ones you have posted.

PJBrs commented:
Hi @bw-mutley, thanks for your tests!

I think two things. First, are you on the [_latest_ version of MSavage's dndtemplate](https://github.com/matsavage/DND-5e-LaTeX-Character-Sheet-Template)? Some of my changes are specifically focused on dealing with recent changes on his side, and they also switched the locations of the attributes and the modifiers.

With regard to my own branch, really, this should be the right command, at the end: 

`git checkout -B pjbrs/MSavage-latex-template`

Verify with 

`git branch`

or 

`git log`

The last commit you should see is MSavage latex template: set XP to milestone if -1


If all else fails, I can send you a tarball with that branch, let me know.

bw-mutley commented:
>P.S., My patches also touch other files than MSavage_template.tex .

Indeed, I should have suspected this to be the case after reading your PR in his repo! I will update with the results as soon as I can test it properly.

PJBrs commented:
@bw-mutley Small heads up: I needed to add one more patch. I just noticed that MSavage's dndtemplate now treats proficiencies a bit differently, which caused the proficiency markers to disappear. My [last patch](https://github.com/PJBrs/dungeon-sheets/commit/d6cfe6bffc9e1dd71cae924fcf90a1cee79ce914) makes them appear again.

You can get that patch with:

```
git fetch pjbrs
git checkout pjbrs/MSavage-latex-template
```

PJBrs commented:
@bw-mutley  ... and I found one final formatting issue in the MSavage latex sheet, so please do one more

```
git fetch pjbrs
git checkout pjbrs/MSavage-latex-template
```
 Last patch now is [MSavage latex template: Prevent spurious newlines](https://github.com/PJBrs/dungeon-sheets/commit/65039930eaf46711b66a328d906141d554a74b50).

I promise that I won't add anything else to my [MSavage-latex-template branch](https://github.com/PJBrs/dungeon-sheets/tree/MSavage-latex-template). It got a little out of hand after I discovered how much MSavage's dndtemplate had progressed, and the extent to which dungeonsheets hadn't kept up with that.

In any case, that branch now is 10 patches ahead of [canismarko/dungeon-sheets:master](https://github.com/canismarko/dungeon-sheets/tree/master), and only touches code that is used in the MSavage latex template. As a check, I just built all examples against it, everything compiles, and also looks fabulous ;-) 

Plan: @bw-mutley to test my MSavage-latex-template branch, and if he approves, I'm turning it into a pull request.

bw-mutley commented:
Hi there, PJ! I will try to check this up carefully in the weekend.

PJBrs commented:
@bw-mutley  OK, thanks for the heads up. In the meantime, I've also had another look at the dndtemplate itself, and @matsavage   and I try to put together a half-caster spellsheet for the latex template. I'll ask him if he might be willing to test. And otherwise, I'll turn it into a new pull request anyway. By now, I've tested my patches with every example in dungeonsheets, and they all render fine!

canismarko commented:
@PJBrs Since you split this off into other PRs, is this still one you're interested in merging?

matsavage commented:
@canismarko this may be due to me not knowing if the half caster sheet would be merged into my repo in time. 

It’s now merged on my end so it may well be both, but I’ll leave for @PJBrs to clarify.

PJBrs commented:
Hi @canismarko 

No, this can definitely not be merged in the current form, but I hope you don't mind using this PR for discussion.

I was thinking about which steps to take next. I think it really helped to snip this PR in smaller pieces. Right now, I'd like to propose the followinf PRs next:

- One with all my current patches affecting @matsavage 's dndtemplate (see this branch: https://github.com/PJBrs/dungeon-sheets/tree/MSavage-latex-template); I think these are completely good to go, I just wanted them to see some testing before raising a PR; I myself compiled all examples with these.
- One with some code refactoring regarding features, with a patches that 1) add class-based darkvision; 2) simplify code in racial spells (implement more racial spells and move them out of race.py); 3) Alert and Remarkable Athlete bonuses to initiative
- One with some ground work in the latex feature pages; the current templates leave many newlines and comments, which doesn't look very clean and it causes some indentation that looks not so nice.

In the meantime, I'm still working on some improvements in my real changes to the feature pages:

1. Ordering spells by level. As you suggested, I have now implemented a command-line option to order spells by level, and I think I should implement associated headings using jinja (my current patch uses latex, but this causes clutter.
2. Ordering features by type. This looks very very nice, but I suspect that you'd like a commandline option for this one as well. Also, my patches don't quite work like I want them to.
3. More fancy decorations, specifically turning dndtables on everywhere in the fancy feature pages (this patch is good) and a patch that uses every available bit of the RPGtex latex style to format monster blocks (this patch makes monsters look very very very nice).
4. Some nice-to-have fixes here and there. 

So, I'll open the first three PR's somewhere this weekend, hopefully after @matsavage has had a look at them, and otherwise - I think I've tested them so often now, they just ought to be good!

And then the half-caster sheet! So I asked @matsavage his opinion about having a nice halfcaster sheet for his template, and we've whipped one up last week, ready for use by dungeonsheets. I haven't implemented that on my side, but it looks rather easy, since we just need to port the halfcaster-sheet code from the reguler fillable pdf to the MSavage code paths. 

Does this work for you? I'll force-push this one here after I update my local branch (point is, I finally discovered whitespace control with jinja2 and how that affects latex output; it makes everything so much better, but now I also know that quite some of my patches were too hacky).

canismarko commented:
@PJBrs No problem at all, just making sure there wasn't anything for me to review right now. Your plan sounds good. For #2, yes I think it makes sense to have it be a command line option, thanks.

PJBrs commented:
I rebased all my branches and created three PRs, as discussed above!

PJBrs commented:
I added two very very very small enhancements to my previous PRs and I finished a sufficiently working patch for ordering feats by type as well. I'll add the MSavage halfcaster sheet next.

@canismarko For now I'll first wait for your reviews of my other three PRs. After that, I can raise a couple more, with 1) more bug fixes,  2) ordering spells by level, and 3) more dnd formatting in the fancy sheets.

PJBrs commented:
... and I implemented the halfcaster sheet for @matsavage 's template! 

@canismarko , for the MSavage latex template I have the patch for portrait and symbol placement in this branch (fixes https://github.com/canismarko/dungeon-sheets/issues/157), and the halfcaster spellsheet implementation. I can either leave them here for a while, or I can add them to my [MSavage latex template PR](https://github.com/canismarko/dungeon-sheets/pull/160). Let me know what you prefer.

Also, I didn't mean to push @Valentin-Metz 's patch that fixes pdfbookmarks when adding images; I had it cherry-picked in my local branch for testing. I can remove it from this PR / branch, but then again, all the attributions are intact, so, I didn't see a problem myself.

PJBrs commented:
I reordered and reorganised the patches in this PR. Together, they now constitute everything that I wanted to add. I added two patches for pdfbookmarks in the character, backstory and spellcasting sheets (both the MSavage template and the fillable pdfs). And I reworded some commit messages to make them easier to understand. All tests pass, all examples build, in latex, in epub. I don't think I can do these patches any better than this. Or at least, not without help :-)

After all this, I can think of only one more "aesthetic" change, and that is switching to lualatex for the MSavage latex template, which is way faster than xelatex, but does add a dependency on the luapstricks latex module.

PJBrs commented:
So, I appear to have messed up the halfcaster patch. It only worked in combination with my order-spells-by-level patch, and even then by chance. I've added a bugfix and will open two more small PRs, one with a couple of fixes and one with a small number of minor enhancements.

In the meantime, it seems to me that one of the showstoppers of using MSavage's dnd template is the problem of the font path, is that right? So, I thought two things; on the one hand, it might be nice to have a fallback font in the dndtemplate.sty file. On the other hand, it shouldn't be too hard to locate the kalam font, with something like the following, but cleaner:
`ls `dirname $(kpsewhich dndtemplate.sty)`/template/fonts/`
And one can then, for instance, softlink the template folder to the current working directory, and clean up afterwards...

PJBrs commented:
The following seems to find the kalam font in the dndtemplate folder:
`TTFONTS=.:$(kpsewhich dndtemplate.sty | sed s/dndtemplate.sty//): makesheets`

PJBrs commented:
So I just succeeded in adding the MSavage dndtemplate.sty itself as a submodule to dungeonsheets, including the right path for the kalam font. It's in my wip branch over at my repo, I'll add it here later. I think that would fix https://github.com/canismarko/dungeon-sheets/issues/92

PJBrs commented:
I added the MSavage latex template as a submodule. I also added it to the tests, and it completes fine. So, I think that, with this, everybody should be able to use the MSavage latex template, regardless of whether they have the kalam font installed or not.

PJBrs commented:
Recently, that is, since I cleaned up my build script and removed the build directory before building dungeonsheets again, I get the following error running makesheets:

```
$ makesheets 
Traceback (most recent call last):
  File "/usr/bin/makesheets", line 5, in <module>
    from dungeonsheets.make_sheets import main
  File "/usr/lib64/python3.9/site-packages/dungeonsheets/__init__.py", line 13, in <module>
    from dungeonsheets import background, features, race, spells, weapons, mechanics
  File "/usr/lib64/python3.9/site-packages/dungeonsheets/mechanics.py", line 4, in <module>
    from dungeonsheets.content import Content, Creature
  File "/usr/lib64/python3.9/site-packages/dungeonsheets/content.py", line 17, in <module>
    __version__ = read("../VERSION").strip()
  File "/usr/lib64/python3.9/site-packages/dungeonsheets/content.py", line 14, in read
    return open((Path(__file__).parent / fname).resolve()).read()
FileNotFoundError: [Errno 2] No such file or directory: '/usr/lib64/python3.9/site-packages/VERSION'
```

This might be fallout from the switchover to pyproject.toml? In any case, it can be fixed with:

```
diff --git a/dungeonsheets/content.py b/dungeonsheets/content.py
index 6fc37ce..1a5e51e 100644
--- a/dungeonsheets/content.py
+++ b/dungeonsheets/content.py
@@ -2,6 +2,8 @@
 
 
 import warnings
+import importlib.metadata
+
 from abc import ABC
 from pathlib import Path
 
@@ -9,13 +11,7 @@ from dungeonsheets import exceptions
 from dungeonsheets.stats import Ability, ArmorClass, Initiative, Speed, Skill
 from dungeonsheets.content_registry import find_content
 
-
-def read(fname):
-    return open((Path(__file__).parent / fname).resolve()).read()
-
-
-__version__ = read("../VERSION").strip()
-
+__version__ = importlib.metadata.version(__package__)
 
 class Content(ABC):
     """A base class for all D&D 5e content types.
```

I can add this to a PR, or create a new one, if you want to.

matsavage commented:
Since this is such a big change to the template, do we think it’s a good time to remove my name from it, since it’s certainly not just me who made the template at this point. 

(It wasn’t me who named it this in the first place btw)

PJBrs commented:
@matsavage What name would you propose? Something like fancy-latex-form.tex?

@canismarko Right now, this PR consists of 24 patches, that cover three types of changes.

First, there are five patches that aim to automatically assign racial spells through their associated feats, with some associated clean-ups (this PR: https://github.com/canismarko/dungeon-sheets/pull/159).

Second, there still are a couple of relatively simple fixes here, fixing the version string for the current build system, a huge patch consisting only of spelling fixes, and a separate spelling fix for sorceror --> sorcererer.

Third, a total of eight patches all aim to make the fancy latex feature pages more beautiful. These commits all start with `Latex feature pages:`. They add: 1) spell sorting by level (and an associated commandline switch), 2) feat sorting by type (and an associated commandline switch), 3) spellcasting details added to the spell spell pages, 4) dndtables for all tables in the fancy sheets and more dndformatting of monsters (this PR: https://github.com/canismarko/dungeon-sheets/pull/165).

Fourth, and finally, I have eight pages that together bring the MSavage latex form (new name pending) at feature parity with the fillable pdf form, in the sense that it contains the same information, doesn't require additional user fiddling by turning it into a submodule, and then some more formatting things as well. These patches may require one or two final tweaks, but I tested them with all examples and they seem to work find.

I'm especially happy with that last series. It would be very nice if it could see some testing. Obviously it works over here, but I didn't test with docker and I didn't test with windows.

matsavage commented:
@PJBrs I'm easy, how about `latex_character_sheet_template.tex`, to differentiate from the pdf templates?

matsavage commented:
All looks good to me, I have confirmed this works with current versions of texlive
* 2024 as a native installation on my machine
* 2022 in the dockerfile, as that is the default which can be installed from the base pythion dockerfile.

I have raised a PR to this branch to update the Dockerfile, as luatex and pstricks are required to be able to build the character sheets in with the fancy template.

I have updated the default makesheets arguments to use the fancy latex sheets, but there is probably some debate to be had about which should be the default, so feel free to revert that.

Note running the docker image with the following command
```
$ docker run -it -v $(pwd):/build ghcr.io/canismarko/dungeon-sheets:master --tex-template
```
will then use the template.

It may be worth updating the README.md to add this information, or maybe to have a section on all the possible output methods, with a sample image of each of them.

There should probably be some debate about which version of the sheet is built by the docker image, or if there should be

matsavage commented:
It may also be worth adding a github actions step to run the docker image against a couple
of the examples with a range of different settings to verify it works before pushing to docker hub.

PJBrs commented:
So, I updated this PR with @matsavage 's patches to update the docker and pyhton-ci workflows. I'll add that I don't really understand docker or github workflows, but it does seem to work, and the docker image now first is tested by compiling all examples and upon success is tagged "latest". So, you can try all patches in this PR by using the docker image that I produced when I pushed to my own repo with:

`docker run -it -v $(pwd):/build ghcr.io/pjbrs/dungeon-sheets:latest  -TSFN`

The commandline switches are as follows: T for --tex-template; S for --spells-by-level; F for --fancy; and N for --feats-by-type.

Do this in a directory that contains your character files. and it ought to work. @matsavage already tested the latex template by using the docker image, and it fully works as intended.

PJBrs commented:
@canismarko How are you doing? I was hoping to get the ball rolling on some PRs again, and in the meantime I also was able to switch the fillable form code from pdfrw to pypdf and to bring it on par with the pdftk code ([in my master branch](https://github.com/PJBrs/dungeon-sheets/tree/master)). This removes the dependency on pdfrw and makes pfdtk fully optional.

I think I fixed a bunch of bugs along the way (the pdfrw actually didn't work in master, and on windows, concatenating pdfs didn't work, if I recall correctly.

matsavage commented:
@PJBrs it may well soon be possible to remove pdftk altogether, I have been doing a bit of work on adding the page flattening to pypdf: https://github.com/py-pdf/pypdf/issues/232#issuecomment-2113335565

PJBrs commented:
I finished the feat sorting work as well, now it does not only sort feats by type, but also subtype (metamagic options under metamagic, etc.). Did a lot of fixes for bloodhunter along the way (four extra patches).

It would be great if this current branch (same as my master branch) sees some testing, especially the feat sorting bits. As mentioned before, testing with docker can be done with:

`docker run -it -v $(pwd):/build ghcr.io/pjbrs/dungeon-sheets:latest -TSFN`

PJBrs commented:
@canismarko I think I'm done for now. I'm sorry that this PR got a bit out of hand, but I encountered all these little issues while trying to implement some features, and then I learned to code better and got a taste for implementing even more, lather, rinse repeat... The last thing I did was go over the current issue list and fix some stuff over there. Oh, and I have a separate branch with some [proficiency refactoring](https://github.com/PJBrs/dungeon-sheets/tree/proficiencies_refactoring) that I've not included here.

Summary:
- Loads of spelling fixes
- Bug fixes for: https://github.com/canismarko/dungeon-sheets/issues/25, https://github.com/canismarko/dungeon-sheets/issues/48, https://github.com/canismarko/dungeon-sheets/issues/150, https://github.com/canismarko/dungeon-sheets/issues/83, https://github.com/canismarko/dungeon-sheets/issues/29, and https://github.com/canismarko/dungeon-sheets/issues/92, as well as 3 bugs that I didn't report.
- Racial spell refactoring and automization
- Maximum DND-style formatting for fancy latex pages
- Commandline option to order spells by level
- Commandline option to order feats by type
- Addition of latex character sheet as git submodule, completely at feature parity with fillable pdf forms
- Removal of pdfrw dependency, and some prep for removing pdftk
- Bloodhunter clean-up

In this post I'll give one more breakdown of all these patches, and I'll signal which I assume are more critical for review.

The following patches already are in a [separate PR](https://github.com/canismarko/dungeon-sheets/pull/166):
97d8e02 Spelling: Sorceror --> Sorcerer
e8258c3 Spellfixes, typos, whitespace fixes
8ee9a81 content.py: Read version from metadata
36ab1ac create_character: Don't crash when no armor is selected
40d2206 create_character: Capitalize all skill proficiencies
2706d65 character.py: Fix adding subclasses
5d7ac60 rogue: Fix spellslots for Arcane Trickster
dfe0d34 weapons: Make unarmed weapons class Unarmed
24d344f create-character: Prevent adding duplicate weapons

These fix seven bugs that I encountered and they include (almost) all spelling errors that I came across.

The following patches are in the [featurestats PR](https://github.com/canismarko/dungeon-sheets/pull/159):
90d2ac4 Features: Classes: Distinguish between racial and class darkvision
1f98530 stats: Include Alert and Remarkable Athlete to initiative
2d664e5 Features: Races: Implement more racial spell features
84ff1ae Race: Move racial spells to features/race
e5adce2 Race: Remove redundant code for spells

Then I have a [third PR](https://github.com/canismarko/dungeon-sheets/pull/165) open that is specifically aimed at achieving maximum DND-style formatting for the fancy latex feature pages while getting the exact functionality that is in the non-fancy latex pages. FOR REVIEW: The DndLongTable environment and the monster formatting jinja filter might seem complicated (but the proof of the pudding is in the eating).
26792d7 Latex feature pages: New DndLongTable environment
b30a2b8 Latex feature pages: Enable DND-tables on fancy sheets
e4ea202 Latex feature pages: Switch fancy gm tables to DndLongTable
b218872 Latex feature pages: Stretch DndLongTable to linewidth
ab49035 Latex feature pages: Use all RPGtex options for monster formatting

Most of the other commits in this PR depend in some way on the above patches (but it wouldn't be very hard to rebase them). The following four patches add a commandline option to order spells by level in the latex feature pages. FOR REVIEW - the first patch in this series adds a short spellcasting section that you may or may not want in there.
c13a781 Latex feature pages: Add spellcasting details to latex spellbook form
1eaa499 Latex feature pages: Sort spells by name and level
b167da3 Latex feature pages: Add subheaders for spell level in the pdf
b910095 README.rst: Document option to order spells by level

The following patches are intended to more logically sort types of features in the latex feature pages, using a commandline option. The first two lay the ground work by separating character feats, class feats, racial feats and background features. The subsequent five patches add advanced feat sorting, so that, for example,  metamagic options are sorted immediately after the metamagic feat and trick shots immediately after the adept marksman feat. FOR REVIEW: I'm assuming here that, when new character feats are added, that they will have `item.source == "Feats"`. Also, I had to do some things like `Rites = CrimsonRites` and `TrickShot = AdeptMarksman`, so that I could sort by checking the parent class of a feat.
58f2685 Latex feature pages: Sort feats by type
fb8f5d1 Latex feature pages: Add headers for feat types
5224cdb Latex feature pages: Advanced feat sorting
7fed1ca Feat sorting: Sort Battle Master maneuvers
d91f27e Feat sorting: Sort TrickShots under AdeptMarksman
95c6183 Feat sorting: Sort Eldritch Invocations
7bd4705 Feat sorting: Bloodhunter Rites, Blood Curses and Formulas
8ef5c0a README.rst: Document feature sorting option

The following patch stands apart a bit. It sets the starting page of the latex feature pages to the next one after the character, personality and spell sheets:
5ffa071 Latex feature pages: Continue page numbering

Then I did a lot of work to bring the latex character sheet by @matsavage at feature parity with the fillable pdf form by adding it as a git submodule. This fixes https://github.com/canismarko/dungeon-sheets/issues/92 . There was a font path issue that blocked further integration of this character sheet, and that's the main point that I fixed. After that, I did some formatting details. Most important in comparison with the fillable forms - I added all classes and levels in the character sheet header, I added overflow spell sheets for when the number of spells exceeds one spell page, and I list spell components on the spell pages, which together bring the latex character sheet exactly at feature parity with the fillable forms. Text will be auto-shrunk to fit with the text boxes on the character sheet. FOR REVIEW: I switched to lualatex, which is a lot faster than xelatex; and I added a patch that uses locally installed dnd latex style and latex character sheet, if it finds them, instead of the copies shipped with dungeon-sheets.
8c86bed dungeonsheets/modules: Add MSavage latex template
d62b22c MSavage latex template: More room for equipment
83115d1 MSavage latex template: Portrait tweaks
0e17889 MSavage latex template: Switch to lualatex
1f8e434 MSavage latex template: List all classes and levels
09debd6 MSavage latex template: List spell requirements
fce6e5e MSavage latex template: Overflow spell sheets
82d28a3 MSavage latex template: Update submodule to v0.1.0
76d2d25 MSavage latex template: Rename
9b7021b README.rst: Document the latex character sheet
dc7086a latex.py: Add local dnd.sty and dndntemplate.sty if installed
b334163 Updates Dockerfile and test pipelines

The last few commits [move towards removing pdftk](https://github.com/canismarko/dungeon-sheets/issues/75), although it is still needed for flattening pdfs. Gets rid of pdfrw (including a bug in the pdfrw code):
80c1edd make_sheets: Use pypdf to merge pdfs when pdftk is absent
34a6e56 fill_pdf_template: switch fallback from pdfrw to pypdf
b0fd311 Forms: Update blank personality sheet

Finally, I noticed a lot of issues with the bloodhunter files that I tried to fix. @plsaddbloodhunter, you might want to have a look at these?  FOR REVIEW: These patches may not be perfect, but I think they fix more problems than they cause. I hope that thes are more or less self-explanatory!
c41fa31 Bloodhunter: 80-column source files and whitespace fixes
c868575 Make bloodhunter example more realistic
2cdb836 Bloodhunter cleanups
1860d56 bloodhunter: integrate duplicate classes

PJBrs commented:
@canismarko I updated this PR again, just to provide a bit of overview. It basically contains four different topics now:

1. Sorting feats by type and subtype ("advanced feat sorting")
2. Adding @matsavage 's latex character template as a proper submodule
3. A load of cleanup of the bloodhunter class code
4. A final five patches to clean up proficiencies

I'll add separate (small) PRs for each of these the coming days.

I'll admit that I have some more patches in my own repo, dealing with latex enhancements and unification, a couple of general code clean-ups, some small enhancements. And that's for later :-)

PJBrs commented:
Decided to close this one.

PJBrs commented:
P.S. Would be nice to discuss on Discord though...
